### DISCUSSION TITLE
Null language id makes rust-analyzer crash

### DISCUSSION THREAD

**Discussion by roife** [joaotavora/eglot#1330] [General]
I have found that rust-analyzer often crashes when used together with eglot. Upon investigation, I discovered that this is because the `languageId` parameter in the `didOpen` request of eglot is set to null, causing rust-analyzer to panic during unwrapping.

Further investigation revealed that this is due to the request issued by `eglot--maybe-activate-editing-mode` in the `after-change-major-mode-hook`. The function `normal-mode` in `lisp/files.el` indicates that `after-change-major-mode-hook` is executed before `set-auto-mode`, causing the current major mode to be `fundamental-mode` at this point. As eglot cannot find the languageId for fundamental mode, it sends null as the languageId. However, according to the LSP spec, the `languageId` parameter must be a String, not String?, so we cannot send null.

In any case, eglot may need to set a default value for languageId, such as "Unknown".

**Comment by joaotavora** [comment]
Analysis is nice, but reproductions is better.  I've reproduced this in one specific situation, using `extend-to-xref`.  Would you mind sharing what you are trying to do following https://joaotavora.github.io/eglot/#Troubleshooting-Eglot?

**Reply by roife** [reply]
Well, here's the event buffer: https://pastebin.mozilla.org/YYbMbQmf.

Every time I use `go-to-definition` and this operation jumps to a file that was not previously open, I encounter this issue. Currently, I am using the following solution to override an existing function to address the problem:

```elisp
(cl-defun eglot--languageId (&optional (server (eglot--current-server-or-lose)))
    "Compute LSP \\='languageId\\=' string for current buffer.
Doubles as a predicate telling if SERVER can manage the current
buffer."
    (or (cl-loop for (mode . languageid) in
                        (eglot--languages server)
                        when (provided-mode-derived-p major-mode mode)
                        return languageid)
	"Unknown"))
```

While this temporarily resolves the issue, the language server reports a duplicate `DidOpenTextDocument` error. I believe this is because the buffer is sending two `didOpen` requests when set to `fundamental-mode` and `rust-mode`.

A more detailed error report would require some time for me to write.

**Reply by joaotavora** [reply]
> A more detailed error report would require some time for me to write.

When you don't spend that time, I will have to spend it in guesswork.  So it's well invested. Let's hope I made the right guess in emacs.git af1fe69f05d803a6958f9d8a045d1013e2ce785c, master branch.  I probably have since according to your event log, your `M-.` to a file outside your project, which was my conjecture, too.  The fix is available from GNU-devel ELPA, if you want to fetch Eglot from there.  There is no duplicate `DidOpen` request in that fix.

**Reply by roife** [reply]
Oh, you are right, I have made some attempts, but there were a few times when I couldn't reproduce it completely, so I am also investigating this issue. I never thought that the reason was that I was jumping into files (libraries) outside of the project. Thank you for your efforts 🙏.
